pyvista/trame/jupyter.py

```diff
@@ -0,0 +1,222 @@
+
+    import pyvista as pv
+    pv.set_jupyter_backend('trame')
+
+If this issue persists, please open an issue in PyVista: https://github.com/pyvista/pyvista/issues
+
+"""
+
+logger = logging.getLogger(__name__)
+
+
+class TrameServerDownError(RuntimeError):
+    """Exception when trame server is down for Jupyter."""
+
+    def __init__(self, server_name):
+        """Call the base class constructor with the custom message."""
+        super().__init__(SERVER_DOWN_MESSAGE.format(name=server_name))
+
+
+class TrameJupyterServerDownError(RuntimeError):
+    """Exception when trame server is down for Jupyter."""
+
+    def __init__(self):
+        """Call the base class constructor with the custom message."""
+        # Be incredibly verbose on how users should launch trame server
+        # Both warn so it appears at top
+        warnings.warn(JUPYTER_SERVER_DOWN_MESSAGE)
+        # and Error
+        super().__init__(JUPYTER_SERVER_DOWN_MESSAGE)
+
+
+def launch_server(server=None):
+    """Launch a trame server."""
+    if server is None:
+        server = pyvista.global_theme.trame.jupyter_server_name
+    if isinstance(server, str):
+        server = get_server(server)
+
+    # Must enable all used modules
+    html_widgets.initialize(server)
+    vtk_widgets.initialize(server)
+    vuetify_widgets.initialize(server)
+
+    def on_ready(**_):
+        logger.debug(f'Server ready: {server}')
+
+    if server._running_stage == 0:
+        server.controller.on_server_ready.add(on_ready)
+        server.start(
+            exec_mode='task',
+            # Default to `127.0.0.1` unless user sets TRAME_DEFAULT_HOST
+            host=os.environ.get('TRAME_DEFAULT_HOST', '127.0.0.1'),
+            port=0,
+            open_browser=False,
+            show_connection_info=False,
+            disable_logging=True,
+            timeout=0,
+        )
+    # else, server is already running
+    return server.ready
+
+
+def build_iframe(_server, ui=None, server_proxy_enabled=None, server_proxy_prefix=None, **kwargs):
+    """Build IPython display.IFrame object for the trame view."""
+    params = f'?ui={ui}&reconnect=auto' if ui else '?reconnect=auto'
+    if server_proxy_enabled is None:
+        server_proxy_enabled = pyvista.global_theme.trame.server_proxy_enabled
+    if server_proxy_enabled:
+        if server_proxy_prefix is None:
+            server_proxy_prefix = pyvista.global_theme.trame.server_proxy_prefix
+        # server_proxy_prefix assumes trailing slash
+        src = (
+            f"{server_proxy_prefix if server_proxy_prefix else ''}{_server.port}/index.html{params}"
+        )
+    else:
+        src = f'{kwargs.get("protocol", "http")}://{kwargs.get("host", "localhost")}:{_server.port}/index.html{params}'
+    logger.debug(src)
+    width = kwargs.get('width', '99%')
+    height = kwargs.get('height', '600px')
+    return widgets.HTML(f"<iframe src='{src}' style='width: {width}; height: {height};'></iframe>")
```

Comment:
If I'm following the code correctly, it seems like there is no longer an easy equivalent of `viewer = plotter.show(..., return_viewer=True); viewer.update_image()`. The function including this line is what gets called under the hood and becomes `viewer` in my code snippet (right?) so it now returns a `widgets.HTML` element, which does not seem to have any options for connecting to the server. Maybe it would be better to return a custom object with a `.server` attribute instead, along with a `_repr_html_` that is the iframe source? Then in MNE where we currently call `viewer.update_image()` to say "rerender/update", we could do `viewer.server.<something>`.

Reply:
I fixed this up. I added a new subclass of the `HTML` `ipywidget` that keeps a reference to the `Viewer` instance, which can be used to sync the widget view. Further, I fixed `return_viewer` to work for the `trame` backend the same way as the other backends.